lib/spack/spack/spec.py

```diff
@@ -1064,8 +1182,8 @@ def __init__(self, spec_like=None,
         self.architecture = None
         self.compiler = None
         self.compiler_flags = FlagMap(self)
-        self._dependents = DependencyMap()
-        self._dependencies = DependencyMap()
+        self._dependents = _EdgeMap(store_by='parent')
+        self._dependencies = _EdgeMap(store_by='child')
```

Comment:
This PR is meant to be part of a more general extension to allow separate concretization of build dependencies. The plan to arrive there is:
1. Refactor specs so that they can model DAGs with the same package present in multiple configurations (this PR)
2. Modify the ASP-based solver to account for cases that currently it can't currently handle, based on the capabilities introduced in 1.

Examples of 2. can be split dependencies in a cross-compilation environment or things like [the one discussed in this thread](https://spackpm.slack.com/archives/C5W7NKZJT/p1620044403296700?thread_ts=1616663239.174500&cid=C5W7NKZJT). I tried to encode in the unit tests 2 cases that will be of interest for future development, see:

https://github.com/spack/spack/blob/83dcaa2cea448bb42c136d4d22679a42a05f6989/lib/spack/spack/test/spec_dag.py#L1002-L1050

TL;DR This PR solves #11983 incidentally, as a result of generalizing the data structures used to keep track of edges in a DAG. It is not meant as a simple fix for just that issue.